Add Standards Positions page on webkit.org

annevk opened the pull request:
#### 6e20509865326ba87283eb645284581141282312
<pre>
Add Standards Positions page on webkit.org
<a href="https://bugs.webkit.org/show_bug.cgi?id=258716">https://bugs.webkit.org/show_bug.cgi?id=258716</a>
rdar://111558028

Reviewed by Alexey Proskuryakov.

This creates a view on top of the summary.json file that I added to the
standards-positions repository so folks have an easier way of browsing
WebKit&apos;s Standards Positions.

* Websites/webkit.org/wp-content/themes/webkit/standards-positions.php: Added.

Canonical link: <a href="https://commits.webkit.org/265929@main">https://commits.webkit.org/265929@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3fbf6cc238a326573e93eb9ce92c2e49cc26db7d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/12311 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/12647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/12957 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/14054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/11850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/15072 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/12664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/14054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/12475 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/15072 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/12957 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/14473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/15072 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/12957 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/14473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/15072 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/12957 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/14473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/11810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/12664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/11033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/12957 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/15364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [  ~~🛠 🧪 unsafe-merge~~](https://ews-build.webkit.org/#/builders/22/builds/1372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/11664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+  <p>Enable JavaScript for an interactive summary table of WebKit's standards positions. Failing that, browse the <a href="https://github.com/WebKit/standards-positions">standards-positions GitHub repository</a> directly.</p>
+</main>
+
+<script>
```
Might be good to run prettier over this file to conform to webkit's style guidelines for JS (at least for 4 char spaces and line length).

annevk replied:
I've used prettier, but it doesn't use 4 spaces it seems.

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+</main>
+
+<script>
+(function () { // Begin closure
```
```suggestion
(async function () { // Begin closure
```

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+  const element = document.createElement(name);
+  for (const name in attributes) {
+    element.setAttribute(name, attributes[name]);
+  }
```
```suggestion
  for (const [name, value] in Object.entries(attributes)) {
    element.setAttribute(name, value);
  }
```

annevk replied:
Also needs s/in/of/. The way this function is called there should never be something on the prototype chain, but I guess we might as well use this pattern.

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+  return parent.appendChild(createElement(name, attributes, child));
+}
+
+function createStandardsPositionsTable() {
```
This seems to do a lot of hard coded work. Why most of the static things as HTML above and then just render in the dynamic aspects. Should make things a bit easier to read/maintain.

annevk replied:
That would probably create a weird experience when JS is off.

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+    }
+    const id = issue["id"].substring("https://github.com/WebKit/standards-positions/issues/".length);
+    
+    const row = addElement(fragment, "tr", { id: `position-${id}` });
```
This really feels like it should be coming from a HTML `<template>` element. You could just clone the structure straight from the template rather than building it up here.

annevk replied:
This is the style we use elsewhere on webkit.org. I'd rather not deviate much from it. And also, `<template>` should have variable support first.

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+  return fragment;
+}
+
+let standardsPositions = null;
```
This being used as a global is a bit confusing in the code... best to pass it as an argument. Similarly with others that are used in the same way. 

These should be defined directly where they are used.

annevk replied:
Passing it as an argument doesn't really work well for a number of cases. Though potentially if we only generate the structure once that could work. I'll see how much work that is.

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+  }
+}
+
+fetch("https://raw.githubusercontent.com/WebKit/standards-positions/main/summary.json").then(res => res.json()).then(positions => {
```
Be good to `await` and `try/catch` for legibility.

annevk replied:
Legibility is in the eye of the beholder, but okay. :-)

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+  }
+}
+
+fetch("https://raw.githubusercontent.com/WebKit/standards-positions/main/summary.json").then(res => res.json()).then(positions => {
```
Thought... maybe as a followup. If we are using PHP anyway, why not just have PHP render tables server side (or embed the JSON inline) and then only include the filtering JS code? It feels like a waste to have every client have to download the JSON file, which doesn't update very often. We could then cache the json response on the server and just have a GitHub hook ping webkit.org any time summary.json gets an update on GitHub.

annevk replied:
Yeah, might make sense, but there are some potential issues around caching as I understand it. This uses roughly the same tactics as are used on `/team/`.

marcoscaceres reviewed (commented):
Exciting stuff.. some suggestions to make this a bit more maintainable.

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+  for (const topic of topics) {
+    addElement(select, "option", {}, topic);
+  }
+}
```
Bit shorter: 

```suggestion
function updateTopicFilter() {
  const topics = new Set([standardsPositions.map(({topics}) => topics).flat().sort()]);
  const select = document.getElementById("filter-topic");
  addElement(select, "option", { value:"" }, "all");
  for (const topic of topics) {
    addElement(select, "option", {}, topic);
  }
}
```

annevk replied:
I find this a lot harder to parse, personally.

marcoscaceres replied:
Hmm, interesting. 

I might just be used to doing things declaratively: with the proposed change one can see declared exactly what's going to happen: `map().flat().sort()` into a unique `Set`. In the current code, everything is all implied, which means actually having to read through it to figure out what it's doing.

I know some folks don't like the "functional programming" style, so could be a matter of preference.

annevk replied:
I think my main problem is that I have a hard time reading the `map()` function and its argument. Whereas a nested for-of loop is very straightforward to read. I'm happy to change it though, not a hill worth dying on.

marcoscaceres replied:
Right, yeah. A lot of people dislike `.map()` and friends for that reason. Would this help make it clear?:

```JS
const toTopics = ({topics}) => topics;
const topics = new Set([standardsPositions.map(toTopics).flat().sort()]);
```

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+function updateStandardsPositionsTableRows() {
+  const filterText = document.getElementById("filter").value;
+  const filterTopic = document.getElementById("filter-topic").value;
+  tbody.replaceChildren(createStandardsPositionsTableRows(standardsPositions, filterText, filterTopic));
```
Just a suggestion, but why not just `<th hidden>` or set a CSS class instead for the ones that don't / do match instead? That way, the JSON structure can go away once the HTML table equivalent is in the DOM.

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,176 @@
+  updateStandardsPositionsTableRows();
+}).catch(() => {
+  loadingChild.replaceWith(createElement("p", {}, "Something went wrong. Please try again later."));
+});
```
Just for us in case something does start going wrong. 

```suggestion
}).catch((err) => {
  loadingChild.replaceWith(createElement("p", {}, "Something went wrong. Please try again later."));
  console.error(err);
});
```

marcoscaceres commented on Websites/webkit.org/wp-content/themes/webkit/standards-positions.php:
```diff
@@ -0,0 +1,266 @@
+<main id=content>
+  <h1><a href="/">Standards Positions</a></h1>
+
+  <p>Enable JavaScript for an interactive summary table of WebKit's standards positions. Failing that, browse the <a href="https://github.com/WebKit/standards-positions">standards-positions GitHub repository</a> directly.</p>
```
Should this be wrapped in `<noscript>`? I would have expected something like:

```HTML
<table>
  ... all the usual table head stuff here disabled....
  <tbody>
    <noscript
      ><tr>
        <td colspan="x">
          <p>
            Enable JavaScript for an interactive summary table of WebKit's
            standards positions. Failing that, browse the
            <a href="https://github.com/WebKit/standards-positions"
              >standards-positions GitHub repository</a
            >
            directly.
          </p>
        </td>
      </tr>
    </noscript>
  </tbody>
</table>

```

annevk replied:
I'm not sure why this is better? It also doesn't parse the way you think it does unfortunately.

marcoscaceres replied:
> It also doesn't parse the way you think it does unfortunately.

yeah, I can see that in the DOM viewer, but it's presented correctly in every browser (i.e., the <noscript> being conditional on the script being disabled):
 
<img width="466" alt="Screenshot 2023-07-07 at 15 42 48" src="https://github.com/WebKit/WebKit/assets/870154/2ecdf056-c5f5-4f55-9acc-1a6a1103e597">

In this sense, every browser is doing exactly what it's supposed to do for this case. That's way better as you don't need to explain what the user is missing out on. They can literally see it...  

> I'm not sure why this is better? 

The reason it's better is that it puts the skeleton of the table back into the markup. Basically, all the JS from line 78-117 in  `createStandardsPositionsTable()` goes away (it's much easier to update the HTML than the JS). 

Consider, is it really necessary to do: 
```JS
 for (const heading of [
    "Specification",
    "Concerns",
    "Position",
    "Topics",
    "Venues",
  ]) {
    addElement(headerRow, "th", { scope: "col" }, heading);
  }
```
  
instead of just this, which is explicitly clear:

```HTML 
<tr>
  <th scope="col">Specification</th>
  <th scope="col">Concerns</th>
  <th scope="col">Position</th>
  <th scope="col">Topics</th>
  <th scope="col">Venues</th>
</tr>
```

Noting also that `addElement()` and `createElement()` adds indirection. 

It's also better because you can then just directly query, populate, and enable the bits of the table based on capability (in this case, JS). 

It's also better from an accessibility perspective because even with JS disabled, the users get to "see" what they are actually not getting (basically what is described in the `<p>`, including all the search input). 

Eventually, if we do populate this table with php, we will need the actual HTML structure there regardless.

Depending how fancy we want to get (and how many people really do hit this site with JS disabled), we could even do filtering serve-side for those folks if we wanted to, while also providing the client side code for local filtering. 

But even if we don't do that, having that can be represented statically/hard-coded not be JS will makes things easier for us to maintain.

marcoscaceres reviewed (approved):
Discussed with @annevk. We will see about moving away from using so much JS in the future if we end up generating this more statically on the server instead.

(I don't have reviewer status, but ✅ from me)

annevk committed:
Add Standards Positions page on webkit.org
https://bugs.webkit.org/show_bug.cgi?id=258716
rdar://111558028

Reviewed by Alexey Proskuryakov.

This creates a view on top of the summary.json file that I added to the
standards-positions repository so folks have an easier way of browsing
WebKit's Standards Positions.

* Websites/webkit.org/wp-content/themes/webkit/standards-positions.php: Added.

Canonical link: https://commits.webkit.org/265929@main
